compiler/codegen/symbol.cpp

```diff
@@ -2227,6 +2268,136 @@ GenRet FnSymbol::codegenPXDType() {
   return ret;
 }
 
+static void pythonRetByteBuffer(FnSymbol* fn, std::string& funcCall,
+                                std::string& returnStmt) {
+
+  // The raw result of the routine call, a "chpl_byte_buffer".
+  funcCall += "cdef chpl_byte_buffer rv = ";
+
+  // Unpack a handle to the buffer and the buffer size.
+  returnStmt += "\tcdef char* rdata = rv.data\n";
+  returnStmt += "\tcdef Py_ssize_t rsize = rv.size\n";
+
+  //
+  // Create a new Python bytes that is a copy of the Chapel bytes. Note
+  // that this call creates a copy of the string.
+  //
+  returnStmt += "\tret = PyBytes_FromStringAndSize(rdata, rsize)\n";
+
+  // This will free the "chpl_byte_buffer" buffer if required.
+  returnStmt += "\tchpl_byte_buffer_free(rv)\n";
+
+  Type* origt = getUnwrappedRetType(fn)->getValType();
+  INT_ASSERT(origt != NULL);
+
+  // The only two types that can be mapped to "chpl_byte_buffer".
+  INT_ASSERT(origt == dtBytes || origt == dtString);
+
+  //
+  // If the original Chapel routine's return type is a string, then
+  // decode the Python bytes into a Python UTF-8 string.
+  //
+  if (origt == dtString) {
+    returnStmt += "\tret = ret.decode(\'utf-8\')\n";  
+  }
+}
+
+static void pythonRetExternalArray(FnSymbol* fn, std::string& funcCall,
+                                   std::string& returnStmt) {
+  INT_ASSERT(fn->retType == dtExternalArray);
+
+  Symbol* eltTypeSym = exportedArrayElementType[fn];
+  if (eltTypeSym == NULL) { return; }
+
+  Type* eltType = eltTypeSym->type;
+
+  funcCall += "cdef chpl_external_array ret_arr = ";
+
+  std::string typeStr = getPythonTypeName(eltType, PYTHON_PYX);
+  std::string typeStrCDefs = getPythonTypeName(eltType, C_PYX);
+
+  //
+  // TODO: Refactor this code and the code in 'pythonRetByteBuffer' so that
+  // I can use both here.
+  //
+  if (eltType == dtBytes || eltType == dtString) {
+
+    std::string res;
+
+    //
+    // Create the numpy array to return. The form looks like:
+    //
+    //  cdef numpy.ndarray [object, ndim=1] ret =
+    //    numpy.zeros(shape = ret_arr.num_elts, dtype = (numpy dtype))
+    //
+    // TODO: Add the 'object' type to the getPythonTypeName table.
+    //
+    res += "\tcdef numpy.ndarray [object, ndim=1] ret = ";
+    res += "numpy.zeros(shape = ret_arr.num_elts, dtype = object)\n";
+
+    //
+    // TODO: Add this to a function we can call with a flag for string or
+    // bytes? Can we also do the same for the opaque array code?
```

Comment:
There's three TODOs here, should they have been resolved already?

Reply:
Dropped down to one TODO.